File: src/main/java/parser/Parser.java
package parser;

import parser.ast.expressions.*;
import parser.ast.statements.*;

import java.util.List;

public final class Parser {

    private static final Token EOF = new Token(TokenType.EOF, "", 0,0);
    private final List<Token> tokens;
    private int pos;
    private final int size;
    public List<String> errors = new CustomArrayList<>();

    public Parser(List<Token> tokens) {
        this.tokens = tokens;
        size = tokens.size();
    }

    public Statement parse() {
        final var result = new BlockStatement();

        while (!match(TokenType.EOF)) {
            result.add(statement());
        }

        return result;
    }

    private Statement block() {
        final var blockStatement = new BlockStatement();
        consume(TokenType.LBRACE, "block");
        while (!match(TokenType.RBRACE)) {
            blockStatement.add(statement());
        }

        return blockStatement;
    }

    private Statement statementOrBlock() {
        if (get(0).getType() == TokenType.LBRACE) {
            return block();
        }

        return statement();
    }

    private Statement statement() {
        if (match(KeywordTokenType.OUT)) {
            return new PrintStatement(expression());
        }

        if (match(KeywordTokenType.IN)) {
            return inputStatement();
        }

        if (match(KeywordTokenType.IF)) {
            return ifElse();
        }

        if (match(KeywordTokenType.MATCH)) {
            return matchStatement();
        }

        if (match(KeywordTokenType.LOOP)) {
            return loopStatement();
        }

        if (match(KeywordTokenType.BREAK)) {
            return new BreakStatement();
        }

        if (match(KeywordTokenType.CONTINUE)) {
            return new ContinueStatement();
        }

        if (match(KeywordTokenType.RET)) {
            return new ReturnStatement(expression());
        }

        if (match(KeywordTokenType.FOR)) {
            return forStatement();
        }

        if (match(KeywordTokenType.FUNC)) {
            return functionDeclaration();
        }

        if (match(KeywordTokenType.CASE)) {
            return caseStatement();
        }

        if (get(0).getType() == TokenType.WORD && get(1).getType() == TokenType.LPAREN) {
            return new FunctionalStatement(function());
        }


        return assignmentStatement();
    }

    private Statement inputStatement() {
        if (get(0).getType() == TokenType.WORD) {
            final var variable = consume(TokenType.WORD, "input statement").getText();
            return new InputStatement(variable);
        }

        throw new RuntimeException("Unknown statement in input statement");
    }

    private Statement assignmentStatement() {
        var tok = get(0);

        if (tok.getType() == TokenType.WORD && get(1).getType() == TokenType.EQ) {
            final var variable = consume(TokenType.WORD, "assignment statement").getText();
            consume(TokenType.EQ, "assignment statement");
            return new AssignmentStatement(variable, expression());
        }

        if (tok.getType() == TokenType.WORD && get(1).getType() == TokenType.LBRACKET) {
            var array = element();
            consume(TokenType.EQ, "assignment statement");
            return new ArrayAssignmentStatement(array, expression());
        }

        if (tok.getType() == TokenType.WORD){
            if (get(1).getType() == TokenType.RPAREN){
                errors.add(String.format("%d:%d: Function \"%s\" doesn't have open paren\n", tok.posfile, tok.posstr, tok.getText()));
                match(TokenType.WORD);
                match(TokenType.RPAREN);
                return new ErrorStatement();
            }
        }

        if (tok.getType() == TokenType.NUMBER && get(1).getType() == TokenType.EQ){
            errors.add(String.format("%d:%d: Unable to assign value to value\n", tok.posfile, tok.posstr));
            match(TokenType.NUMBER);
            match(TokenType.EQ);
            match(TokenType.NUMBER);
            return new ErrorStatement();
        }

        if (get(0).getType() == TokenType.LBRACE) {
            return statementOrBlock();
        }

        throw new RuntimeException("Unknown statement near!");
    }

    private Statement caseStatement() {
        final var conditionVariant = expression();
        consume(TokenType.ARROW, "case statement");
        final var body = statementOrBlock();
        return new CaseStatement(conditionVariant, body);
    }

    private Statement matchStatement() {
        final var condition = expression();
        final var matchBody = statementOrBlock();
        return new MatchStatement(condition, matchBody);
    }


    private Statement ifElse() {
        final var condition = expression();
        final var ifStatement = statementOrBlock();
        final Statement elseStatement;

        if (match(KeywordTokenType.ELSE)) {
            elseStatement = statementOrBlock();
        } else {
            elseStatement = null;
        }

        return new IfStatement(condition, ifStatement, elseStatement);
    }

    private Statement loopStatement() {
        final var condition = expression();
        final var statement = statementOrBlock();
        return new LoopStatement(condition, statement);
    }

    private Statement forStatement() {
        final var init = assignmentStatement();
        consume(TokenType.SEMI_COLON, "for statement");
        final var term = expression();
        consume(TokenType.SEMI_COLON, "for statement");
        final var inc = assignmentStatement();
        final var stat = statementOrBlock();
        return new ForStatement(init, term, inc, stat);
    }


    private FunctionDeclarationStatement functionDeclaration() {
        final var name = consume(TokenType.WORD, "function declaration").getText();
        consume(TokenType.LPAREN, "function declaration");
        final List<String> argNames = new CustomArrayList<>();
        while (!match(TokenType.RPAREN)) {
            argNames.add(consume(TokenType.WORD, "function declaration").getText());
            match(TokenType.COMMA);
        }
        final var instructions = statementOrBlock();
        return new FunctionDeclarationStatement(name, argNames, instructions);
    }


    private FunctionalExpression function() {
        final var name = consume(TokenType.WORD, "function expression").getText();
        consume(TokenType.LPAREN, "function expression");
        final var function = new FunctionalExpression(name);
        while (!match(TokenType.RPAREN)) {
            var tok = get(0);
            if (tok.getType() instanceof KeywordTokenType){
                errors.add(String.format("%d:%d: Function \"%s\" doesn't have close paren\n", tok.posfile, tok.posstr, name));
                break;
            }
            function.addArgument(expression());
            match(TokenType.COMMA);
        }
        return function;
    }

    private Expression expression() {
        return logicalOr();
    }

    private Expression logicalOr() {
        var result = logicalAnd();

        while (true) {
            if (match(TokenType.BARBAR)) {
                result = new ConditionalExpression(ConditionalExpression.Operator.OR, result, logicalAnd());
                continue;
            }
            break;
        }

        return result;
    }

    private Expression logicalAnd() {
        var result = equality();
        while (true) {
            if (match(TokenType.AMPAMP)) {
                result = new ConditionalExpression(ConditionalExpression.Operator.AND, result, equality());
                continue;
            }
            break;
        }
        return result;
    }

    private Expression equality() {
        var result = conditional();

        if (match(ComparisonTokenType.EQEQ)) {
            return new ConditionalExpression(ConditionalExpression.Operator.EQUALS, result, conditional());
        }

        if (match(ComparisonTokenType.EXCLEQ)) {
            System.out.println("Da");
            return new ConditionalExpression(ConditionalExpression.Operator.NOT_EQUALS, result, conditional());
        }

        return result;
    }

    private Expression conditional() {
        var result = additive();
        while (true) {
            if (match(ComparisonTokenType.LT)) {
                result = new ConditionalExpression(ConditionalExpression.Operator.LT, result, additive());
                continue;
            }

            if (match(ComparisonTokenType.LTEQ)) {
                result = new ConditionalExpression(ConditionalExpression.Operator.LTEQ, result, additive());
                continue;
            }

            if (match(ComparisonTokenType.GT)) {
                result = new ConditionalExpression(ConditionalExpression.Operator.GT, result, additive());
                continue;
            }

            if (match(ComparisonTokenType.GTEQ)) {
                result = new ConditionalExpression(ConditionalExpression.Operator.GTEQ, result, additive());
                continue;
            }

            break;
        }

        return result;
    }

    private Expression additive() {
        var result = multiplicative();
        while (true) {
            final var token = get(0);
            if (match(ArithmeticTokenType.PLUS)) {
                result = new BinaryExpression('+', result, multiplicative(), token.posstr, token.posfile);
                continue;
            }

            if (match(ArithmeticTokenType.MINUS)) {
                result = new BinaryExpression('-', result, multiplicative(), token.posstr, token.posfile);
                continue;
            }
            break;
        }

        return result;
    }

    private Expression multiplicative() {
        var result = unary();
        while (true) {
            final var token = get(0);
            if (match(ArithmeticTokenType.STAR)) {
                result = new BinaryExpression('*', result, unary(), token.posstr, token.posfile);
                continue;
            }

            if (match(ArithmeticTokenType.SLASH)) {
                result = new BinaryExpression('/', result, unary(), token.posstr, token.posfile);
                continue;
            }
            break;
        }

        return result;
    }

    private Expression unary() {
        if (match(ArithmeticTokenType.MINUS)) {
            return new UnaryExpression('-', primary());
        }

        if (match(ArithmeticTokenType.PLUS)) {
            return primary();
        }

        return primary();
    }

    private Expression primary() {
        final var current = get(0);
        if (match(TokenType.NUMBER)) {
            return new ValueExpression(Double.parseDouble(current.getText()));
        }

        if (get(0).getType() == TokenType.WORD && get(1).getType() == TokenType.LPAREN) {
            return function();
        }

        if (get(0).getType() == TokenType.WORD && get(1).getType() == TokenType.LBRACKET) {
            return element();
        }

        if (get(0).getType() == TokenType.LBRACKET) {
            return array();
        }

        if (match(TokenType.WORD)) {
            return new VariableExpression(current.getText());
        }

        if (match(TokenType.TEXT)) {
            return new ValueExpression(current.getText());
        }

        if (match(TokenType.LPAREN)) {
            var result = expression();
            if (match(TokenType.RPAREN)){
                return result;
            }
        }

        errors.add(String.format("%d:%d: Don't assign the value near \"%s\" \n", current.posfile, current.posstr, current.getText()));
        return new ErrorExpression();
        //throw new RuntimeException();
    }

    private Expression array() {
        consume(TokenType.LBRACKET, "array expression");
        final List<Expression> elements = new CustomArrayList<>();
        while (!match(TokenType.RBRACKET)) {
            elements.add(expression());
            match(TokenType.COMMA);
        }
        return new ArrayExpression(elements);
    }

    private ArrayAccessExpression element() {
        final var variable = consume(TokenType.WORD, "array access expression").getText();
        List<Expression> indexes = new CustomArrayList<>();
        do {
            consume(TokenType.LBRACKET, "array access expression");
            indexes.add(expression());
            consume(TokenType.RBRACKET, "array access expression");
        } while (get(0).getType() == TokenType.LBRACKET);

        final var prevToken = get(-2);
        return new ArrayAccessExpression(variable, indexes, prevToken.posstr, prevToken.posfile);
    }

    private Token consume(TokenType type, String place) {
        final var current = get(0);
        if (type != current.getType()) {
            errors.add(String.format("%d:%d: %s doesn't match type %s in %s", current.posfile, current.posstr, current, type, place));
            return null;
        }
        pos++;
        return current;
    }

    private boolean match(ITokenType type) {
        final var current = get(0);
        if (type != current.getType())
            return false;

        pos++;
        return true;
    }

    private Token get(int relativePosition) {
        final var position = pos + relativePosition;

        if (position >= size)
            return EOF;

        return tokens.get(position);
    }
}

File: src/main/java/parser/Parser.kt
package parser

import parser.ast.expressions.*
import parser.ast.statements.*

class Parser(private val tokens: List<Token>) {
    private var pos: Int = 0
    private val size: Int = tokens.size
    var errors: MutableList<String> = CustomArrayList()

    fun parse(): Statement {
        val result = BlockStatement()

        while (!match(TokenType.EOF)) result.add(statement())

        return result
    }

    private fun block(): Statement {
        val blockStatement = BlockStatement()
        consume(TokenType.LBRACE, "block")
        while (!match(TokenType.RBRACE)) blockStatement.add(statement())

        return blockStatement
    }

    private fun statementOrBlock(): Statement = when {
        get(0).type === TokenType.LBRACE -> block()
        else -> statement()
    }

    private fun statement(): Statement {
        when {
            match(KeywordTokenType.OUT) -> return PrintStatement(expression())
            match(KeywordTokenType.IN) -> return inputStatement()
            match(KeywordTokenType.IF) -> return ifElse()
            match(KeywordTokenType.MATCH) -> return matchStatement()
            match(KeywordTokenType.LOOP) -> return loopStatement()
            match(KeywordTokenType.BREAK) -> return BreakStatement()
            match(KeywordTokenType.CONTINUE) -> return ContinueStatement()
            match(KeywordTokenType.RET) -> return ReturnStatement(expression())
            match(KeywordTokenType.FOR) -> return forStatement()
            match(KeywordTokenType.FUNC) -> return functionDeclaration()
            match(KeywordTokenType.CASE) -> return caseStatement()
            else -> return when {
                get(0).type === TokenType.WORD && get(1).type === TokenType.LPAREN -> FunctionalStatement(function())
                else -> assignmentStatement()
            }
        }


    }

    private fun inputStatement(): Statement {
        if (get(0).type === TokenType.WORD) {
            val variable = consume(TokenType.WORD, "input statement")!!.text
            return InputStatement(variable)
        }

        throw RuntimeException("Unknown statement in input statement")
    }

    private fun assignmentStatement(): Statement {
        val tok = get(0)

        if (tok.type === TokenType.WORD && get(1).type === TokenType.EQ) {
            val variable = consume(TokenType.WORD, "assignment statement")!!.text
            consume(TokenType.EQ, "assignment statement")
            return AssignmentStatement(variable, expression())
        }

        if (tok.type === TokenType.WORD && get(1).type === TokenType.LBRACKET) {
            val array = element()
            consume(TokenType.EQ, "assignment statement")
            return ArrayAssignmentStatement(array, expression())
        }

        when {
            tok.type === TokenType.WORD -> if (get(1).type === TokenType.RPAREN) {
                errors.add(String.format("%d:%d: Function \"%s\" doesn't have open paren\n", tok.posfile, tok.posstr, tok.text))
                match(TokenType.WORD)
                match(TokenType.RPAREN)
                return ErrorStatement()
            }
        }

        if (tok.type === TokenType.NUMBER && get(1).type === TokenType.EQ) {
            errors.add(String.format("%d:%d: Unable to assign value to value\n", tok.posfile, tok.posstr))
            match(TokenType.NUMBER)
            match(TokenType.EQ)
            match(TokenType.NUMBER)
            return ErrorStatement()
        }

        if (get(0).type === TokenType.LBRACE) {
            return statementOrBlock()
        }

        throw RuntimeException("Unknown statement near!")
    }

    private fun caseStatement(): Statement {
        val conditionVariant = expression()
        consume(TokenType.ARROW, "case statement")
        val body = statementOrBlock()
        return CaseStatement(conditionVariant, body)
    }

    private fun matchStatement(): Statement {
        val condition = expression()
        val matchBody = statementOrBlock()
        return MatchStatement(condition, matchBody)
    }


    private fun ifElse(): Statement {
        val condition = expression()
        val ifStatement = statementOrBlock()
        val elseStatement: Statement?

        when {
            match(KeywordTokenType.ELSE) -> elseStatement = statementOrBlock()
            else -> elseStatement = null
        }

        return IfStatement(condition, ifStatement, elseStatement)
    }

    private fun loopStatement(): Statement {
        val condition = expression()
        val statement = statementOrBlock()
        return LoopStatement(condition, statement)
    }

    private fun forStatement(): Statement {
        val init = assignmentStatement()
        consume(TokenType.SEMI_COLON, "for statement")
        val term = expression()
        consume(TokenType.SEMI_COLON, "for statement")
        val inc = assignmentStatement()
        val stat = statementOrBlock()
        return ForStatement(init, term, inc, stat)
    }


    private fun functionDeclaration(): FunctionDeclarationStatement {
        val name = consume(TokenType.WORD, "function declaration")!!.text
        consume(TokenType.LPAREN, "function declaration")
        val argNames = CustomArrayList<String>()
        while (!match(TokenType.RPAREN)) {
            argNames.add(consume(TokenType.WORD, "function declaration")!!.text)
            match(TokenType.COMMA)
        }
        val instructions = statementOrBlock()
        return FunctionDeclarationStatement(name, argNames, instructions)
    }


    private fun function(): FunctionalExpression {
        val name = consume(TokenType.WORD, "function expression")!!.text
        consume(TokenType.LPAREN, "function expression")
        val function = FunctionalExpression(name)
        while (!match(TokenType.RPAREN)) {
            val tok = get(0)
            if (tok.type is KeywordTokenType) {
                errors.add(String.format("%d:%d: Function \"%s\" doesn't have close paren\n", tok.posfile, tok.posstr, name))
                break
            }
            function.addArgument(expression())
            match(TokenType.COMMA)
        }
        return function
    }

    private fun expression(): Expression {
        return logicalOr()
    }

    private fun logicalOr(): Expression {
        var result = logicalAnd()

        while (true) {
            if (match(TokenType.BARBAR)) {
                result = ConditionalExpression(ConditionalExpression.Operator.OR, result, logicalAnd())
                continue
            }
            break
        }

        return result
    }

    private fun logicalAnd(): Expression {
        var result = equality()
        while (true) {
            if (match(TokenType.AMPAMP)) {
                result = ConditionalExpression(ConditionalExpression.Operator.AND, result, equality())
                continue
            }
            break
        }
        return result
    }

    private fun equality(): Expression {
        val result = conditional()

        return when {
            match(ComparisonTokenType.EQEQ) -> ConditionalExpression(ConditionalExpression.Operator.EQUALS, result, conditional())
            match(ComparisonTokenType.EXCLEQ) -> {
                println("Da")
                ConditionalExpression(ConditionalExpression.Operator.NOT_EQUALS, result, conditional())
            }
            else -> result
        }

    }

    private fun conditional(): Expression {
        var result = additive()
        while (true) {
            if (match(ComparisonTokenType.LT)) {
                result = ConditionalExpression(ConditionalExpression.Operator.LT, result, additive())
                continue
            }

            if (match(ComparisonTokenType.LTEQ)) {
                result = ConditionalExpression(ConditionalExpression.Operator.LTEQ, result, additive())
                continue
            }

            if (match(ComparisonTokenType.GT)) {
                result = ConditionalExpression(ConditionalExpression.Operator.GT, result, additive())
                continue
            }

            if (match(ComparisonTokenType.GTEQ)) {
                result = ConditionalExpression(ConditionalExpression.Operator.GTEQ, result, additive())
                continue
            }

            break
        }

        return result
    }

    private fun additive(): Expression {
        var result = multiplicative()
        while (true) {
            val token = get(0)
            if (match(ArithmeticTokenType.PLUS)) {
                result = BinaryExpression('+', result, multiplicative(), token.posstr, token.posfile)
                continue
            }

            if (match(ArithmeticTokenType.MINUS)) {
                result = BinaryExpression('-', result, multiplicative(), token.posstr, token.posfile)
                continue
            }
            break
        }

        return result
    }

    private fun multiplicative(): Expression {
        var result = unary()
        while (true) {
            val token = get(0)
            if (match(ArithmeticTokenType.STAR)) {
                result = BinaryExpression('*', result, unary(), token.posstr, token.posfile)
                continue
            }

            if (match(ArithmeticTokenType.SLASH)) {
                result = BinaryExpression('/', result, unary(), token.posstr, token.posfile)
                continue
            }
            break
        }

        return result
    }

    private fun unary(): Expression {
        return when {
            match(ArithmeticTokenType.MINUS) -> UnaryExpression('-', primary())
            else -> when {
                match(ArithmeticTokenType.PLUS) -> primary()
                else -> primary()
            }
        }

    }

    private fun primary(): Expression {
        val current = get(0)
        if (match(TokenType.NUMBER)) {
            return ValueExpression(java.lang.Double.parseDouble(current.text))
        }

        if (get(0).type === TokenType.WORD && get(1).type === TokenType.LPAREN) {
            return function()
        }

        if (get(0).type === TokenType.WORD && get(1).type === TokenType.LBRACKET) {
            return element()
        }

        if (get(0).type === TokenType.LBRACKET) {
            return array()
        }

        if (match(TokenType.WORD)) {
            return VariableExpression(current.text)
        }

        if (match(TokenType.TEXT)) {
            return ValueExpression(current.text)
        }

        if (match(TokenType.LPAREN)) {
            val result = expression()
            if (match(TokenType.RPAREN)) {
                return result
            }
        }

        errors.add(String.format("%d:%d: Don't assign the value near \"%s\" \n", current.posfile, current.posstr, current.text))
        return ErrorExpression()
    }

    private fun array(): Expression {
        consume(TokenType.LBRACKET, "array expression")
        val elements = CustomArrayList<Expression>()
        while (!match(TokenType.RBRACKET)) {
            elements.add(expression())
            match(TokenType.COMMA)
        }
        return ArrayExpression(elements)
    }

    private fun element(): ArrayAccessExpression {
        val variable = consume(TokenType.WORD, "array access expression")!!.text
        val indexes = CustomArrayList<Expression>()
        do {
            consume(TokenType.LBRACKET, "array access expression")
            indexes.add(expression())
            consume(TokenType.RBRACKET, "array access expression")
        } while (get(0).type === TokenType.LBRACKET)

        val prevToken = get(-2)
        return ArrayAccessExpression(variable, indexes, prevToken.posstr, prevToken.posfile)
    }

    private fun consume(type: TokenType, place: String): Token? {
        val current = get(0)
        if (type !== current.type) {
            errors.add(String.format("%d:%d: %s doesn't match type %s in %s", current.posfile, current.posstr, current, type, place))
            return null
        }
        pos++
        return current
    }

    private fun match(type: ITokenType): Boolean {
        val current = get(0)
        if (type !== current.type)
            return false

        pos++
        return true
    }

    private operator fun get(relativePosition: Int): Token {
        val position = pos + relativePosition

        return if (position >= size) EOF else tokens[position]

    }

    companion object {

        private val EOF = Token(TokenType.EOF, "", 0, 0)
    }
}
